src/lib.rs

```diff
@@ -153,21 +208,135 @@ pub struct Graph<'a> {
 impl<'a> Graph<'a> {
     /// Creates a new [`Graph`] object to be partitioned.
     ///
+    /// - `ncon` is the number of constraints on each vertex (at least 1),
+    /// - `nparts` is the number of parts wanted in the graph partition.
+    ///
+    /// # Input format
+    ///
+    /// `xadj` and `adjncy` are the [CSR encoding][0] of the adjacency matrix
+    /// that represents the graph. `xadj` is the row index and `adjncy` is the
+    /// column index.
+    ///
+    /// [0]: https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_(CSR,_CRS_or_Yale_format)
+    ///
+    /// # Errors
+    ///
+    /// The following invariants must be held, otherwise this function returns
+    /// an error:
+    ///
+    /// - all of the arrays have a length that can be held by an [`Idx`],
+    /// - `ncon` is strictly greater than zero,
+    /// - `nparts` is strictly greater than zero,
+    /// - `xadj` has at least one element (its length is the one more than the
+    ///   number of vertices),
+    /// - `xadj` is sorted,
+    /// - elements of `xadj` are positive,
+    /// - the last element of `xadj` is the length of `adjncy`,
+    /// - elements of `adjncy` are within zero and the number of vertices.
+    ///
+    /// # Mutability
+    ///
+    /// [`Graph::part_kway`] and [`Graph::part_recursive`] may mutate the
+    /// contents of `xadj` and `adjncy`, but should revert all changes before
+    /// returning.
+    pub fn new(
+        ncon: Idx,
+        nparts: Idx,
+        xadj: &'a mut [Idx],
+        adjncy: &'a mut [Idx],
+    ) -> std::result::Result<Graph<'a>, NewError> {
+        if ncon <= 0 {
+            return Err(NewError {
+                kind: NewErrorKind::InvalidNumberOfConstraints,
+            });
+        }
+        if nparts <= 0 {
```

Comment:
```suggestion
        if nparts <= 1 {
```
Same deal as with `Mesh::new` - this test fails with SIGFPE
```rust
// using old function signatures
#[test]
fn mesh_sigfpe() {
    let adjncy = &mut [1, 0];
    let xadj = &mut [0, 1, 2];

    let mut part = [0; 2];

    // passing npart == 1 raises SIGFPE
    Graph::new(1, 1, xadj, adjncy)
        .part_kway(&mut part)
        .unwrap();
}
```